diff --git a/pm_mac/pmmacosxcm.c b/pm_mac/pmmacosxcm.c
index e8b196c..7e48bca 100755
--- a/pm_mac/pmmacosxcm.c
+++ b/pm_mac/pmmacosxcm.c
@@ -156,7 +156,9 @@
    timestamped messages a little ahead of real time, not for
    scheduling an entire MIDI sequence at once!
  */
+#ifndef LIMIT_RATE
 #define LIMIT_RATE 1
+#endif
 
 #define SYSEX_BUFFER_SIZE 128
 /* What is the maximum PortMidi device number for an IAC device? A
